server/commands/admin.py

```diff
@@ -133,16 +137,90 @@ def ooc_cmd_ban(client, arg):
     Usage: /ban <ipid> "reason" ["<N> <minute|hour|day|week|month>(s)|perma"]
     Usage 2: /ban <ipid> <ban_id>
     """
-    kickban(client, arg, False)
-
+    kickban(client, arg, BanOptions(include_hdid=False))
 
+@mod_only()
 def ooc_cmd_banhdid(client, arg):
     """
     Ban both a user's HDID and IPID.
-    DANGER: Banning webAO users by HDID has unintended consequences.
     Usage: See /ban.
     """
-    kickban(client, arg, True)
+    kickban(client, arg, BanOptions(include_hdid=True))
+
+def _find_area(client, area_name):
+    try:
+        return client.server.area_manager.get_area_by_id(int(area_name))
+    except:
+        try:
+            return client.server.area_manager.get_area_by_name(area_name)
+        except ValueError:
+            raise ArgumentError('Area ID must be a name or a number.')
+
+def ooc_cmd_area_curse(client, arg):
+    """
+    Ban a player from all areas except one, such that when they connect, they
+    will be placed in a specified area and can't switch areas unless forcefully done
+    by a moderator.
+
+    To unban, use the /unban command.
+    To add more IPIDs/HDIDs, use the /ban command as usual.
+
+    Usage: /area_curse <ipid> <area_name> "reason" ["<N> <minute|hour|day|week|month>(s)|perma"]
+    """
+    args = shlex.split(arg)
+    default_ban_duration = client.server.config['default_ban_duration']
+
+    if len(args) < 3:
+        raise ArgumentError('Not enough arguments.')
+    else:
+        ipid = _convert_ipid_to_int(args[0])
+        target_area = _find_area(client, args[1])
+        reason = args[2]
+
+    if len(args) == 3:
+        ban_duration = parse(str(default_ban_duration))
+        unban_date = arrow.get().shift(seconds=ban_duration).datetime
+    elif len(args) == 4:
+        duration = args[3]
+        ban_duration = parse(str(duration))
+
+        if duration is None:
+            raise ArgumentError('Invalid ban duration.')
+        elif 'perma' in duration.lower():
+            unban_date = None
+        else:
+            if ban_duration is not None:
+                unban_date = arrow.get().shift(seconds=ban_duration).datetime
+            else:
+                raise ArgumentError(f'{duration} is an invalid ban duration')
+
+    else:
+        raise ArgumentError(f'Ambiguous input: {arg}\nPlease wrap your arguments '
+                            'in quotes.')
+
+    special_ban_data = json.dumps({
+        'ban_type': 'area_curse',
+        'target_area': target_area.id
+    })
+
+    ban_id = database.ban(ipid, reason, ban_type='ipid', banned_by=client,
+                          unban_date=unban_date, special_ban_data=special_ban_data)
+
+    targets = client.server.client_manager.get_targets(
+        client, TargetType.IPID, ipid, False)
+
+    for c in targets:
+        c.send_ooc(f'You were area cursed by {client.name}.')
```

Comment:
i am not so sure if it is a smart move to tell the victim who did it

Reply:
k